lib/shared/errors.js

```diff
@@ -0,0 +1,366 @@
+ * check a property called `report`. For backwards compatibility, and
+ * to properly handle errors that we don't control, an error without
+ * this property is reported automatically.
+ *
+ * @param {Error} error - error
+ * @returns {Boolean} whether the error should be reported
+ *
+ * @example
+ * if (errors.shouldReport(new Error('foo'))) {
+ *   console.log('We should report this error');
+ * }
+ */
+exports.shouldReport = (error) => {
+  return !_.has(error, 'report') || Boolean(error.report);
+};
+
+/**
+ * @summary Check if a string is blank
+ * @function
+ * @private
+ *
+ * @param {String} string - string
+ * @returns {Boolean} whether the string is blank
+ *
+ * @example
+ * if (isBlank('   ')) {
+ *   console.log('The string is blank');
+ * }
+ */
+const isBlank = _.flow([ _.trim, _.isEmpty ]);
+
+/**
+ * @summary Get the title of an error
+ * @function
+ * @public
+ *
+ * @description
+ * Try to get as most information as possible about the error
+ * rather than falling back to generic messages right away.
+ *
+ * @param {Error} error - error
+ * @returns {String} error title
+ *
+ * @example
+ * const error = new Error('Foo bar');
+ * const title = errors.getTitle(error);
+ * console.log(title);
+ */
+exports.getTitle = (error) => {
+  if (!_.isError(error) && !_.isPlainObject(error) && !_.isNil(error)) {
+    return _.toString(error);
+  }
+
+  const codeTitle = getUserFriendlyMessageProperty(error, 'title');
+  if (!_.isNil(codeTitle)) {
+    return codeTitle;
+  }
+
+  const message = _.get(error, 'message');
+  if (!isBlank(message)) {
+    return message;
+  }
+
+  const code = _.get(error, 'code');
+  if (!_.isNil(code) && !isBlank(code)) {
+    return `Error code: ${code}`;
+  }
+
+  return 'An error ocurred';
+};
+
+/**
+ * @summary Get the description of an error
+ * @function
+ * @public
+ *
+ * @param {Error} error - error
+ * @returns {String} error description
+ *
+ * @example
+ * const error = new Error('Foo bar');
+ * const description = errors.getDescription(error);
+ * console.log(description);
+ */
+exports.getDescription = (error) => {
+  if (!_.isError(error) && !_.isPlainObject(error)) {
+    return '';
+  }
+
+  if (!isBlank(error.description)) {
+    return error.description;
+  }
+
+  const codeDescription = getUserFriendlyMessageProperty(error, 'description');
+  if (!_.isNil(codeDescription)) {
+    return codeDescription;
+  }
+
+  if (error.stack) {
+    return error.stack;
+  }
+
+  if (_.isEmpty(error)) {
+    return '';
+  }
+
+  return JSON.stringify(error, null, 2);
+};
+
+/**
+ * @summary Create an error
+ * @function
+ * @public
+ *
+ * @param {String} title - error title
+ * @param {String} [description] - error description
+ * @param {Object} [options] - options
+ * @param {Boolean} [options.report] - report error
+ * @returns {Error} error
+ *
+ * @example
+ * const error = errors.createError('Foo', 'Bar');
+ * throw error;
+ */
+exports.createError = (title, description, options = {}) => {
```

Comment:
The description is an optional argument (see the JSDoc annotation above), and it feels that its overkill for some cases (e.g: where you just want to throw a real Etcher error).